### PR TITLE
feat(BA-3625): Add UUID Filter for strawberry gql object

### LINKED ISSUE #7685: Add ID filter for Strawberry GQL Objects
## Motivation

Currently, Strawberry GQL objects do not have a standardized filter for ID/UUID-based fields. While filtering by other field types (string, int, datetime) is supported through `StringFilter`, `IntFilter`, etc., there's no reusable filter for ID fields. This leads to inconsistent filtering approaches across different GQL objects.

‌

## Required Features

1. **Create IDFilter input type in base.py**
   - Add new `IDFilter` input type to `src/ai/backend/manager/api/gql/base.py`
   - Follow the same pattern as existing filters (`StringFilter`, `IntFilter`, `DateTimeFilter`)
   - Include the following filter operations:
     - `equals: Optional[str]` - Exact match
     - `not_equals: Optional[str]` (GraphQL name: `notEquals`) - Not equal
     - `in_: Optional[list[str]]` (GraphQL name: `in`) - Match any of the provided IDs
     - `not_in: Optional[list[str]]` (GraphQL name: `notIn`) - Not match any of the provided IDs
   - Implement `build_query_condition()` method that takes factory callables:
     - `equals_factory: Callable[[str], QueryCondition]` - For equals/not_equals operations
     - `in_factory: Callable[[list[str]], QueryCondition]` - For in/not_in operations
   - Handle NOT operations using `~` operator for `not_equals` and `not_in`
1. **Update ScalingGroupFilterGQL to use IDFilter**
   - Replace proposed single `id` field with `id: Optional[IDFilter]` in `ScalingGroupFilterGQL`
   - Update `build_conditions()` to use `IDFilter.build_query_condition()`
   - Ensure proper integration with existing AND/OR/NOT logical operators
1. **Implement ID condition factories in ScalingGroupConditions**
   - Add `by_id(id: str) -> QueryCondition` method for equals operation
   - Add `by_ids(ids: list[str]) -> QueryCondition` method for in operation
   - These will internally use name-based lookup since ScalingGroup's ID maps to name
1. **Make IDFilter reusable across codebase**
   - Design IDFilter as a generic, reusable component
   - Other GQL objects (ModelDeployment, ModelRevision, etc.) can adopt this filter
   - Promotes consistency across the GraphQL API

‌

## Impact

**Components affected:**

- `src/ai/backend/manager/api/gql/base.py`
  - **NEW**: Add `IDFilter` input type with `build_query_condition()` method
  - Follow existing filter patterns for consistency
- `src/ai/backend/manager/api/gql/scaling_group/types.py`
  - `ScalingGroupFilterGQL` class - Add `id: Optional[IDFilter]` field and update `build_conditions()` method
- `src/ai/backend/manager/repositories/scaling_group/options.py`
  - `ScalingGroupConditions` class - Add `by_id()` and `by_ids()` static methods
- **Services affected:**
- No service layer changes required (filtering logic is handled at repository level)
**Database schema:**
- No schema changes required (using existing `name` column which is already the primary key)
**API changes:**
- **Breaking change**: None (adding optional filter field is backward compatible)
- **New capability**:
  - Clients can filter by ID with multiple operations (equals, not_equals, in, not_in)
  - Consistent filter pattern across all GQL objects
  - Future GQL objects can reuse IDFilter for their ID-based filtering needs

‌

## Testing Scenarios

1. **IDFilter operations**
   - Test `equals` with valid ID (should return single matching item)
   - Test `not_equals` with valid ID (should return all except specified ID)
   - Test `in` with multiple valid IDs (should return all matching items)
   - Test `not_in` with multiple IDs (should return all except specified IDs)
   - Test with non-existent IDs (should return empty results or all items depending on operation)
1. **IDFilter edge cases**
   - Test with empty list in `in` operation (should return empty result)
   - Test with empty list in `not_in` operation (should return all items)
   - Test with single item in `in` operation (equivalent to `equals`)
   - Test with null/None values (should be ignored)
1. **ID filter with other filters**
   - Test combining `id` filter with `is_active` filter using AND operator
   - Test combining `id` filter with name contains filter
   - Verify logical operators (AND/OR/NOT) work correctly with ID filter
   - Test precedence: only one operation should be applied if multiple fields are set
1. **Integration with ScalingGroup queries**
   - Test `scaling_groups_v2` query with ID filter and project parameter
   - Test `all_scaling_groups_v2` query with ID filter
   - Verify pagination works correctly with ID filtering
   - Test ordering with ID filtering
1. **GraphQL API validation**
   - Verify GraphQL schema correctly exposes `IDFilter` input type
   - Test query introspection returns correct filter structure
   - Verify field names are properly mapped (e.g., `in_` → `in`, `not_equals` → `notEquals`)
   - Verify filter description appears in schema documentation
1. **QueryCondition integration**
   - Verify `~` operator correctly negates conditions for `not_equals` and `not_in`
   - Test that generated SQL queries are correct and optimized
   - Verify proper escaping of special characters in IDs
1. **Reusability validation**
   - Verify IDFilter can be imported and used in other GQL object filters
   - Test consistency of behavior when used across different objects

## Implementation Notes

- Use `str` type for ID fields as GraphQL ID scalar serializes to string
- Use `strawberry.field(name="...")` to map Python field names to GraphQL schema names
- Follow operation precedence: `equals` > `not_equals` > `in` > `not_in`
- Ensure factory callables properly handle both positive and negative conditions
- Add comprehensive docstrings explaining the filter behavior

JIRA Issue: BA-3625

### PR DESCRIPTION
**Pull request by seedspirit**
resolves #7685 (BA-3625)

**Checklist:** (if applicable)

- [ ] Milestone metadata specifying the target backport version
- [x] Mention to the original issue
- [ ] Installer updates including:
  - Fixtures for db schema changes
  - New mandatory config options
- [ ] Update of end-to-end CLI integration tests in `ai.backend.test`
- [ ] API server-client counterparts (e.g., manager API -> client SDK)
- [ ] Test case(s) to:
  - Demonstrate the difference of before/after
  - Demonstrate the flow of abstract/conceptual models with a concrete implementation
- [ ] Documentation
  - Contents in the `docs` directory
  - docstrings in public interfaces and type annotations

### CODE REVIEW

File: src/ai/backend/manager/api/gql/base.py [hunk 1]
```diff
@@ -115,6 +115,47 @@ def to_dataclass(self) -> IntFilterData:
         )
 
 
+@strawberry.input(description="Added in 26.1.0. Filter for UUID fields.")
+class UUIDFilter:
+    equals: Optional[uuid.UUID] = None
+    not_equals: Optional[uuid.UUID] = strawberry.field(name="notEquals", default=None)
```
Comment (HyeockJinKim) [inline src/ai/backend/manager/api/gql/base.py]: ~Why wasn't `equals` applied in strawberry.field?~ Ignore this.

File: src/ai/backend/manager/api/gql/base.py [hunk 2]
```diff
@@ -115,6 +115,47 @@ def to_dataclass(self) -> IntFilterData:
         )
 
 
+@strawberry.input(description="Added in 26.1.0. Filter for UUID fields.")
+class UUIDFilter:
```
Comment (HyeockJinKim) [inline src/ai/backend/manager/api/gql/base.py:223]: Could you inherit the GQLFilter class?

File: src/ai/backend/manager/api/gql/base.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (HyeockJinKim) [reply src/ai/backend/manager/api/gql/base.py]: It doesn't need to be applied to `not_equals`.

File: src/ai/backend/manager/api/gql/base.py [hunk 3]
```diff
@@ -115,6 +115,47 @@ def to_dataclass(self) -> IntFilterData:
...
+        """
+        if self.equals:
+            return equals_factory(self.equals)
+        elif self.not_equals:
+            return self._negate_condition(equals_factory(self.not_equals))
+        elif self.in_:
+            return in_factory(self.in_)
+        elif self.not_in:
+            return self._negate_condition(in_factory(self.not_in))
+        return None
```
Comment (HyeockJinKim) [inline src/ai/backend/manager/api/gql/base.py:278]: Could you take a look at this, referencing the StringFilter? There's a risk of confusion if the pattern is only different from this one.

File: src/ai/backend/manager/api/gql/base.py [hunk 4]
```diff
@@ -203,6 +203,75 @@ def to_dataclass(self) -> IntFilterData:
...
+    # Basic operations
+    equals: str | None = None
+    in_: list[str] | None = strawberry.field(name="in", default=None)
+
+    # NOT operations
+    not_equals: str | None = None
+    not_in: list[str] | None = None
+    # Case-insensitive operations
+    i_equals: str | None = None
+    i_in: list[str] | None = None
```
Comment (HyeockJinKim) [inline src/ai/backend/manager/api/gql/base.py]: `i_in` doesn't seem necessary. We can just use `not_in` if needed.

Comment (HyeockJinKim) [reply src/ai/backend/manager/api/gql/base.py]: StringMatchSpec is designed for Strings, not UUIDs. Please define UUIDMatchSpec to only accept the value and whether it's not a UUID.

File: src/ai/backend/manager/api/gql/base.py [hunk 5]
```diff
@@ -33,6 +33,16 @@ class StringMatchSpec:
     negated: bool
 
 
+@dataclass(frozen=True)
+class UUIDMatchSpec:
+    """Specification for UUID matching operations in query conditions."""
+
+    value: uuid.UUID | list[uuid.UUID]
+    case_insensitive: bool
```
Comment (fregataa) [inline src/ai/backend/manager/api/gql/base.py]: UUID values are always case sensitive

File: src/ai/backend/manager/api/gql/base.py [hunk 6]
```diff
@@ -33,6 +33,15 @@ class StringMatchSpec:
     negated: bool
 
 
+@dataclass(frozen=True)
+class UUIDMatchSpec:
+    """Specification for UUID matching operations in query conditions."""
+
+    value: uuid.UUID | list[uuid.UUID]
+    negated: bool
+    is_list_operation: bool = False
```
Comment (HyeockJinKim) [inline src/ai/backend/manager/api/gql/base.py]: We receive equals and in separately... I don't want to use a Union type.

**Review by HyeockJinKim — CHANGES_REQUESTED** [review]
We should add the implementation for the UUID filter to the Repositories side.
